### ISSUE TITLE
Possible OOM Condition Encountered? If So Can Source Code Be Modified to Prevent OOM Error From Happening?

### ISSUE THREAD

**Issue by tedliosu** [CNugteren/CLBlast#444]
```
(base) tedliosu@victus-ted:~/Documents/all_git/CLBlast_vs_CUDA/build$ python3 ../scripts/benchmark/benchmark.py --comparisons CPU-BLAS cuBLAS --platform 0 --device 0 --precision 6464 --benchmark gemm
[benchmark] Running on platform 0, device 0
[benchmark] Running 6 benchmarks for settings 'gemm'
[benchmark] Running benchmark 'gemm:multiples of 128'
[benchmark] Calling binary: ./clblast_client_xgemm -platform 0 -device 0 -precision 6464 -runs 20 -warm_up -q -no_abbrv -clblas 0 -cblas 1 -cublas 1 -m 128 -n 128 -k 128 -layout 102 -transA 111 -transB 111 -step 128 -num_steps 20
[benchmark] Running benchmark 'gemm:multiples of 129'
[benchmark] Calling binary: ./clblast_client_xgemm -platform 0 -device 0 -precision 6464 -runs 20 -warm_up -q -no_abbrv -clblas 0 -cblas 1 -cublas 1 -m 129 -n 129 -k 129 -layout 102 -transA 111 -transB 111 -step 129 -num_steps 20
terminate called after throwing an instance of 'std::runtime_error'
  what():  CUDA cudaMalloc error with status: 2
Aborted (core dumped)
```
I highly suspect that the error was caused by an OOM condition because when I ran the same benchmark with the parameter `--precision 64` instead (i.e. a dgemm instead of a zgemm benchmark), VRAM usage at one point shot up to > 4000 MB:
![vram_usage_benchmark](https://user-images.githubusercontent.com/25697946/194697158-fecefe00-8636-4219-a479-2a072f08dea4.png)
(As you can see I have a 3050 mobile so I only have 4 GB VRAM).

With that being said, is it possible for buffers to be deallocated at some point during the benchmark with a modification of the source code so that I may be able to run the zgemm benchmark on my 3050 mobile that compares CLBlast to _both_ cuBLAS and CPU-BLAS? I really do wish that I got a laptop with more VRAM except that I was on a budget and Nvidia products with larger amounts of VRAM tend to be much more expensive :confused:

**Comment by CNugteren** [comment]
I looked at the code of the benchmark 'client' again and it looks like it allocates and de-allocates at each step. So running this command:
```
./clblast_client_xgemm -platform 0 -device 0 -precision 6464 -runs 20 -warm_up -q -no_abbrv -clblas 0 -cblas 1 -cublas 1 -m 129 -n 129 -k 129 -layout 102 -transA 111 -transB 111 -step 129 -num_steps 20
```
Means that at each of the 20 steps there will be allocations and de-allocations. In this case the largest dimension will be 2580 (=20 * 129), and thus given that there are 3 matrixes and each element is 16 bytes we get:
```
(2580 * 2580 * 3 * 16) / (1024. * 1024.) = around 300MB
```
So that is a lot lower than your 4GB. Internally BLAS algorithms might use a bit extra memory for optimizations, but that is still not close to the 4GB. You do have both the reference (cuBLAS in this case) and CLBlast running, so you'll have to multiple the number already by 2.

In short: I'm not sure why this doesn't fit on your device, but I don't think I can help much. Try reproducing the issue in a stand-alone call to `clblast_client_xgemm` and try to find out which sizes work and which ones don't. E.g. start with something as simple as `./clblast_client_xgemm -m 4096 -n 4096 -k 4096` and do the size math (as above) yourself. Then explore other sizes and/or enable cuBLAS.

But really if you just want to run the benchmarks I would just lower the sizes till they fit, e.g. change the `20` here into `15` or so for starters: https://github.com/CNugteren/CLBlast/blob/master/scripts/benchmark/settings.py#L153

**Comment by tedliosu** [comment]
Per @CNugteren's advice, I reduced the appropriate numbers in the GEMM section in settings.py until there were no more OOM errors, and I discovered that for some reason OOM errors would start occuring when matrix sizes for ZGEMM operations were around 2100 x 2100 x 2100 or greater, so I capped the matrix sizes at around 2060 x 2060 x 2060, as you may see in the attached patch file here: [settings-py-patch.txt](https://github.com/CNugteren/CLBlast/files/9849667/settings-py-patch.txt).  So hopefully the modified settings.py after applying the patch will serve as an acceptable workaround for those who end up running into similar issues in the future as I did, esp. as dGPUs with as little as 4 GB of VRAM were still being released as recently as [earlier this year](https://www.techpowerup.com/gpu-specs/radeon-rx-6500-xt.c3850).

I am closing this issue for now as I don't really have to time to debug the root of the cause of the DGEMM and ZGEMM benchmarks taking up _way_ more memory than expected.
